elements/pfe-clipboard/test/pfe-clipboard.spec.ts

```diff
@@ -0,0 +1,211 @@
+    // Default text should be set the feedback--success variable
+    expect(
+      getComputedStyle(el.shadowRoot!.querySelector(`.pfe-clipboard__text--success`)!)
+        .getPropertyValue('color'),
+      'success style'
+    ).to.equal(`rgb(${hexToRgb('#3e8635').join(', ')})`);
+  });
+
+  it('should fire a deprecated pfe-clipboard:copied event when clicked', async function() {
+    const handlerSpy = spy();
+    document.body.addEventListener('pfe-clipboard:copied', handlerSpy);
+    const el = await createFixture<PfeClipboard>(element);
+    // Add global event listener for the copy event
+    // Simulate click
+    setTimeout(el.click.bind(el));
+    const event = await oneEvent(document.body, 'pfe-clipboard:copied');
+    expect(event.detail.url).to.be.ok;
+    expect(event.detail.copiedText).to.be.ok;
+    await aTimeout(100);
+    expect(handlerSpy).to.have.been.calledOnce;
+    document.body.removeEventListener('pfe-clipboard:copied', handlerSpy);
+  });
+
+  it('should fire a copied event when clicked', async function() {
+    const handlerSpy = spy();
+    document.body.addEventListener('copied', handlerSpy);
+    const el = await createFixture<PfeClipboard>(element);
+    // Add global event listener for the copy event
+    // Simulate click
+    setTimeout(el.click.bind(el));
+    const event = await oneEvent(document.body, 'copied');
+    expect(event.detail.copiedText).to.not.be.empty;
+    await aTimeout(100);
+    expect(handlerSpy).to.have.been.calledOnce;
+    document.body.removeEventListener('pfe-clipboard:copied', handlerSpy);
+  });
+
+  it(`should have the correct accessibility attributes`, async function() {
+    const el = await createFixture<PfeClipboard>(element);
+    // Add global event listener for the copy event
+    expect(el.getAttribute('role')).to.equal('button');
+    expect(el.getAttribute('tabindex')).to.equal('0');
+  });
+
+  it(`should support copying the url by default`, async function() {
+    const el = await createFixture<PfeClipboard>(element);
+    // click the element
+    el.click();
+    expect(navigator.clipboard.writeText).to.have.been.calledWith(location.href);
+  });
+
+  it(`should support copying arbitrary text using contentToCopy property`, async function() {
+    const el = await createFixture<PfeClipboard>(element);
+    const copyText = `<div>Copy this text</div>`;
+    // manually set the contentToCopy property
+    el.copyFrom = 'property';
+    await el.updateComplete;
+    expect(
+      el.hasAttribute('disabled'),
+      'pfe-clipboard should be disabled when target=property but "contentToCopy" has not been set'
+    ).to.be.true;
+    el.contentToCopy = copyText;
+    await el.updateComplete;
+    expect(
+      el.hasAttribute('disabled'),
+      'pfe-clipboard should not be disabled when target=property and "contentToCopy" has been set'
+    ).to.be.false;
+    el.click();
+    expect(navigator.clipboard.writeText).to.have.been.calledWith(copyText);
```

Comment:
This is cool!  So we get `calledWith` from Sinon's `stubs` function?

Reply:
yeah it's got quite some beef to it

https://www.chaijs.com/plugins/sinon-chai/

sinon-chai lags behind sinon a bit so you can always just get a reference to the stub if you want some of the more advanced apis

Reply:
🎉